File: snap_prep/0350.cpp
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        if(nums1.size() > nums2.size()){
            return intersect(nums2, nums1);
        }
        
        unordered_map<int, int> dic;
        
        for(int n : nums1){
            ++dic[n];
        }
        
        int i = 0;
        for(int n : nums2){
            if(dic.find(n) != dic.end()){
                nums1[i++] = n;
                dic[n]--;
                if(dic[n] == 0){
                    dic.erase(n);
                }
            }
        }
        
        return vector(begin(nums1), begin(nums1) + i);
    }
};

File: snap_prep/0350.java
class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        if(nums1.length > nums2.length){
            return intersect(nums2, nums1);
        }
        
        HashMap<Integer, Integer> map = new HashMap<>();
        
        for(int n : nums1){
            int val = map.getOrDefault(n, 0);
            map.put(n, val + 1);
        }
        
        int i = 0;
        for(int n : nums2){
            if(map.containsKey(n)){
                nums1[i++] = n;
                int val = map.get(n) - 1;
                if(val > 0){
                    map.put(n, val);
                }else{
                    map.remove(n);
                }
            }
        }
        
        return Arrays.copyOfRange(nums1, 0, i);
    }
}
